# snATAC-seq
## Pipeline for Split & Pool snATAC-seq Data Processing to Generate Fragment Files
Note: process each demultiplexed sample separately

### Requirements

This pipeline requires the following software:

- **BWA**: >=0.7.17
- **samtools**: >=1.12
- **snaptools**: >=1.4.8

### Additional Files

To run this pipeline, you also need to download the following files provided in this GitHub repository:

- **BarcodeCorrect.py**: Script for extracting cell barcode from read 1 and generate a barcode dictionary with up to 1 mismatch.
- **FragmentCorrect.py**: Script for correcting barcodes of fragments.
- **scATAC_v2_barcode_list.txt.gz**: Barcode whitelist file.

Make sure to download these files and place them in the appropriate directory before running the pipeline.

### Usage

```bash
#!/bin/bash

threads=10 # Specify the number of threads to use
sample="sample1" # the prefix for the FASTQ files
BWA_index="PATH_TO_YOUR/mm10.fa" # the path of bwa index for reference genome
gsize="PATH_TO_YOUR/mm10.chrom.sizes" # the path of genome size file
tmpfold="PATH_TO_YOUR/temp" # temporary directory
logfile="PATH_TO_YOUR/snaptools.log"
outbam="PATH_TO_YOUR/${sample}.bam"
outsnap="PATH_TO_YOUR/${sample}.snap"
output="PATH_TO_YOUR/${sample}.temp.tsv.gz"
corrected_output="PATH_TO_YOUR/${sample}.tsv"
fragment_output="PATH_TO_YOUR/${sample}.tsv.gz"
fastq1="PATH_TO_YOUR/${sample}_R1.fastq.gz"
fastq2="PATH_TO_YOUR/${sample}_R2.fastq.gz"

# Step 1: Align paired-end reads
snaptools align-paired-end --input-reference=${BWA_index} --input-fastq1=${fastq1} --input-fastq2=${fastq2} --output-bam=${outbam} --aligner=bwa --read-fastq-command=zcat --min-cov=0 --num-threads=${threads} --if-sort=True --tmp-folder=${tmpfold} --overwrite=TRUE

# Step 2: Generate SNAP file from BAM
snaptools snap-pre --input-file=${outbam} --output-snap=${outsnap} --genome-name=mm10 --genome-size=${gsize} --min-mapq=10 --min-flen=0 --keep-chrm=TRUE --keep-single=False --keep-secondary=False --overwrite=True --max-num=10000000 --verbose=False

# Step 3: Generate fragment file from SNAP file
snaptools dump-fragment --snap-file=${outsnap} --output-file=${output} --buffer-size=10000 --tmp-folder=${tmpfold} &> ${logfile}

# Step 4: Extract cell barcode from read 1 and generate a barcode dictionary with up to 1 mismatch
python BarcodeCorrect.py  --fq ${fastq1}  -b scATAC_v2_barcode_list.txt.gz -O barcode_correct.txt

# Step 5: Correct fragment file barcodes
python FragmentCorrect.py -F ${output} -C barcode_correct.txt -O ${corrected_output}

# Step 6: Adjust Tn5 insertion offsets
awk '{print$1,$2+4,$3-5,$4}' OFS='\t' ${corrected_output} | sort -V -k1,1 -k2,2n -T ${tmpfold} |  pbgzip -c > ${fragment_output}

# Step 7: Index the corrected fragment file
tabix -p bed ${fragment_output}

# Optional: Clean up intermediate files
#rm -f ${output} barcode_correct.txt ${corrected_output}
```

### Output

The final fragment file `${sample}.tsv.gz` generated by this pipeline can be used as an input file for **ArchR** ([https://github.com/GreenleafLab/ArchR](https://github.com/GreenleafLab/ArchR)) to perform downstream analyses.
